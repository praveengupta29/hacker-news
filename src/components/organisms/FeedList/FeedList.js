import React from 'react';
import PropTypes from 'prop-types';
import FeedItem from '../../molecules/FeedItem';

import {
  filterFeedDataWithHiddenFeeds,
  updateUpVoteInFeedData,
} from '../../../containers/organisms/FeedList/utils';
import './FeedList.scss';

const FeedList = ({ feedListData }) => {
  const filteredData = filterFeedDataWithHiddenFeeds(feedListData);
  const votedData = updateUpVoteInFeedData(filteredData);

  const { hits: feedList } = votedData;

  return (
    <div className="feed-list-container">
      {feedList.map(feed => {
        const { created_at_i: timestamp } = feed;

        return <FeedItem {...feed} key={timestamp} />;
      })}
    </div>
  );
};

FeedList.propTypes = {
  feedListData: PropTypes.object.isRequired,
};

export default FeedList;
