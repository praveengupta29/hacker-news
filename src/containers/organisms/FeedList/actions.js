import {
  FETCH_FEED_LIST_DATA,
  FETCH_FEED_LIST_DATA_SUCCESS,
  FETCH_FEED_LIST_DATA_ERROR,
} from './constants';

export const fetchFeedListData = params => ({
  type: FETCH_FEED_LIST_DATA,
  params,
});

export const fetchFeedListDataSuccess = data => ({
  type: FETCH_FEED_LIST_DATA_SUCCESS,
  data,
});

export const fetchFeedListDataError = error => ({
  type: FETCH_FEED_LIST_DATA_ERROR,
  error,
});
