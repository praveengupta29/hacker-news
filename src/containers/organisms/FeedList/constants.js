export const FETCH_FEED_LIST_DATA =
  'HACKER_NEWS/FEED_LIST/FETCH_FEED_LIST_DATA';
export const FETCH_FEED_LIST_DATA_SUCCESS =
  'HACKER_NEWS/FEED_LIST/FETCH_FEED_LIST_DATA_SUCCESS';
export const FETCH_FEED_LIST_DATA_ERROR =
  'HACKER_NEWS/FEED_LIST/FETCH_FEED_LIST_DATA_ERROR';
export const UP_VOTE_FEED = 'HACKER_NEWS/FEED_LIST/UP_VOTE_FEED';
