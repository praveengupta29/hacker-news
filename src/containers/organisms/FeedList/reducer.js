import {
  FETCH_FEED_LIST_DATA,
  FETCH_FEED_LIST_DATA_SUCCESS,
  FETCH_FEED_LIST_DATA_ERROR,
} from './constants';

import feedListInitialState from '../../../commons/initialState/feedList.state';

const initialState = feedListInitialState;

const feedListReducer = (state = initialState, { type, data, error }) => {
  switch (type) {
    case FETCH_FEED_LIST_DATA:
      return {
        ...state,
        isFetching: true,
      };

    case FETCH_FEED_LIST_DATA_SUCCESS:
      return {
        ...state,
        isFetching: false,
        data,
      };

    case FETCH_FEED_LIST_DATA_ERROR:
      return {
        ...state,
        isFetching: false,
        error,
      };

    default:
      return state;
  }
};

export default feedListReducer;
