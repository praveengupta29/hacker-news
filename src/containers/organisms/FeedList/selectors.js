import { createSelector } from 'reselect';

const getFeedListDataState = state => {
  return state.feedList;
};

export const selectFeedListData = createSelector(
  getFeedListDataState,
  feedListDataState => feedListDataState.data,
);
