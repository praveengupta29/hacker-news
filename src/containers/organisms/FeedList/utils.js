import { LocalStorageUtil } from '../../../utils/localStorage';
import { HIDE_FEED_KEY } from '../../atoms/HideFeed/constants';

export const removeFeed = (feedListState, objectID) => {
  const { data, ...others } = feedListState;
  const { hits } = data;

  const feedIndex = hits.findIndex(feed => {
    return feed.objectID === objectID;
  });

  hits.splice(feedIndex, 1);

  return {
    data: {
      ...data,
      hits: [...hits],
    },
    ...others,
  };
};

export const filterFeedDataWithHiddenFeeds = feedListData => {
  const storage = new LocalStorageUtil();

  const hiddenFeedIds = JSON.parse(storage.getItem(HIDE_FEED_KEY) || '[]');

  if (hiddenFeedIds.length) {
    const { hits } = feedListData;
    const filteredHits = hits.filter(hit => {
      return hiddenFeedIds.indexOf(hit.objectID) === -1;
    });

    return {
      ...feedListData,
      hits: filteredHits,
    };
  }
  return feedListData;
};
