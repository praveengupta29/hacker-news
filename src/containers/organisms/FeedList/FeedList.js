import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';

import { selectFeedListData } from './selectors';
import { fetchFeedListData } from './actions';

import FeedList from '../../../components/organisms/FeedList';

export const mapStateToProps = createStructuredSelector({
  feedListData: selectFeedListData,
});

export const mapDispatchToProps = dispatch => {
  return {
    fetchFeedListData: () => {
      dispatch(fetchFeedListData());
    },
  };
};

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(FeedList);
