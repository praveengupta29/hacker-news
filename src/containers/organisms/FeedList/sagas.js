import { call, put, takeLatest } from 'redux-saga/effects';
import history from '../../../utils/history';

import { fetchFeedListDataSuccess, fetchFeedListDataError } from './actions';

import { FETCH_FEED_LIST_DATA } from './constants';

import { FRONT_PAGE_TAG } from '../../../commons/constants';
import apis from '../../../commons/constants/api.services';

import ServiceUtil from '../../../commons/utils/ServiceUtil';
import { buildUrl } from '../../../commons/utils/url';

const updateQueryParams = params => {
  if (history && Object.keys(params).indexOf('page') > -1) {
    const queryParamsUrl = buildUrl({
      pathName: '/',
      query: {
        ...params,
      },
    });

    history.push(queryParamsUrl);
  }
};

export function* loadFeedListSaga(action) {
  try {
    const { params = {} } = action;

    const url = buildUrl({
      pathName: apis.search,
      query: {
        tags: FRONT_PAGE_TAG,
        ...params,
      },
    });

    const listingData = yield call(ServiceUtil.fetch, url);

    yield put(fetchFeedListDataSuccess(listingData));

    updateQueryParams(params);
  } catch (err) {
    yield put(fetchFeedListDataError(err));
  }
}

export default function* flightListingSaga() {
  yield takeLatest(FETCH_FEED_LIST_DATA, loadFeedListSaga);
}
