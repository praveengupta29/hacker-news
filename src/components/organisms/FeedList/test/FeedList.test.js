import React from 'react';
import { shallow } from 'enzyme';
import FeedList from '../FeedList';
import feedListData from '../../../../containers/organisms/FeedList/tests/feeds.mock.json';

describe('FeedList component', () => {
  it('should render correctly and match snapshot', () => {
    const component = shallow(<FeedList feedListData={feedListData} />);

    expect(component).toMatchSnapshot();
  });
});
