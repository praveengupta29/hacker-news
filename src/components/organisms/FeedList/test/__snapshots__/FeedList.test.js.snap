// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeedList component should render correctly and match snapshot 1`] = `
<div
  className="feed-list-container"
>
  <FeedItem
    _highlightResult={
      Object {
        "author": Object {
          "matchLevel": "none",
          "matchedWords": Array [],
          "value": "astdb",
        },
        "title": Object {
          "matchLevel": "none",
          "matchedWords": Array [],
          "value": "Alan Kay: ““Maxwell’s Equations of Software ”“",
        },
        "url": Object {
          "matchLevel": "none",
          "matchedWords": Array [],
          "value": "https://queue.acm.org/detail.cfm?id=1039523",
        },
      }
    }
    _tags={
      Array [
        "story",
        "author_astdb",
        "story_22847079",
        "front_page",
      ]
    }
    author="astdb"
    comment_text={null}
    created_at="2020-04-12T09:54:21.000Z"
    created_at_i={1586685261}
    key="1586685261"
    num_comments={4}
    objectID="22847079"
    parent_id={null}
    points={34}
    story_id={null}
    story_text={null}
    story_title={null}
    story_url={null}
    title="Alan Kay: ““Maxwell’s Equations of Software ”“"
    url="https://queue.acm.org/detail.cfm?id=1039523"
  />
  <FeedItem
    _highlightResult={
      Object {
        "author": Object {
          "matchLevel": "none",
          "matchedWords": Array [],
          "value": "kiyanwang",
        },
        "title": Object {
          "matchLevel": "none",
          "matchedWords": Array [],
          "value": "Technology-budgeting/handbook.md at master · 18F/technology-budgeting",
        },
        "url": Object {
          "matchLevel": "none",
          "matchedWords": Array [],
          "value": "https://github.com/18F/technology-budgeting/blob/master/handbook.md",
        },
      }
    }
    _tags={
      Array [
        "story",
        "author_kiyanwang",
        "story_22847877",
        "front_page",
      ]
    }
    author="kiyanwang"
    comment_text={null}
    created_at="2020-04-12T12:44:02.000Z"
    created_at_i={1586695442}
    key="1586695442"
    num_comments={4}
    objectID="22847877"
    parent_id={null}
    points={27}
    story_id={null}
    story_text={null}
    story_title={null}
    story_url={null}
    title="Technology-budgeting/handbook.md at master · 18F/technology-budgeting"
    url="https://github.com/18F/technology-budgeting/blob/master/handbook.md"
  />
</div>
`;
